gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1426 @@
+    if (zoneContainer == null) {
+      return Error.NOT_FOUND.response(String.format(
+          "The 'parameters.managedZone' resource named '%s' does not exist.", zoneName));
+    }
+    List<RrsetWrapper> dnsRecords = zoneContainer.dnsRecords().get(zoneName);
+    String[] fields = (String[]) options.get("fields");
+    String name = (String) options.get("name");
+    String type = (String) options.get("type");
+    String pageToken = (String) options.get("pageToken");
+    Integer maxResults = options.get("maxResults") == null
+        ? null : Integer.valueOf((String) options.get("maxResults"));
+    boolean listing = (pageToken == null); // matches will be included in the result if true
+    boolean sizeReached = false; // maximum result size was reached, we should not return more
+    boolean hasMorePages = false; // should next page token be included in the response?
+    LinkedList<String> serializedRrsets = new LinkedList<>();
+    String lastRecordId = null;
+    for (RrsetWrapper recordWrapper : dnsRecords) {
+      ResourceRecordSet record = recordWrapper.rrset();
+      if (listing) {
+        if (matchesCriteria(record, name, type)) {
+          if (sizeReached) {
+            // we do not add this, just note that there would be more and there should be a token
+            hasMorePages = true;
+            break;
+          } else {
+            lastRecordId = recordWrapper.id();
+            try {
+              serializedRrsets.addLast(jsonFactory.toString(
+                  OptionParsersAndExtractors.extractFields(record, fields)));
+            } catch (IOException e) {
+              return Error.INTERNAL_ERROR.response(String.format(
+                  "Error when serializing resource record set in managed zone %s in project %s",
+                  zoneName, projectId));
+            }
+          }
+        }
+      }
+      // either we are listing already, or we check if we should start in the next iteration
+      listing = recordWrapper.id().equals(pageToken) || listing;
+      sizeReached = (maxResults != null) && maxResults.equals(serializedRrsets.size());
+    }
+    boolean includePageToken =
+        hasMorePages && (fields == null || ImmutableList.copyOf(fields).contains("nextPageToken"));
+    return toListResponse(serializedRrsets, "rrsets", lastRecordId, includePageToken);
+  }
+
+  /**
+   * Lists changes. Next page token is ID of the last change listed.
+   */
+  Response listChanges(String projectId, String zoneName, Map<String, Object> options) {
+    Response response = checkListOptions(options);
+    if (response != null) {
+      return response;
+    }
+    ZoneContainer zoneContainer = findZone(projectId, zoneName);
+    if (zoneContainer == null) {
+      return Error.NOT_FOUND.response(String.format(
+          "The 'parameters.managedZone' resource named '%s' does not exist", zoneName));
+    }
+    // take a sorted snapshot of the current change list
```

Comment:
What's the reasoning behind avoiding the ConcurrentLinkedQueue's iterator instead?